src/avram/model.cr

```diff
@@ -184,6 +184,44 @@ class Avram::Model
     {% COLUMNS << {name: type_declaration.var, type: data_type, nilable: nilable.id, autogenerated: autogenerated} %}
   end
 
+  macro polymorphic_belongs_to(type_declaration, autogenerated = false, optional = false)
+
+    {{ nilable = !optional }}
+
+    {% COLUMNS << {name: type_declaration.var + "_id", type: Int64, nilable: nilable.id, autogenerated: autogenerated} %}
+    {% COLUMNS << {name: type_declaration.var + "_type", type: String, nilable: nilable.id, autogenerated: autogenerated} %}
+
```

Comment:
To validate_inclusion_of you can add a before_save in here:

```crystal
macro finished
  class SaveOperation
    before_save :validate_polymorphic_association
       
    private def validate_polymorphic_association
      validate_inclusion_of {{ type_declartion.var }}_type, in: []
    end
  end
end
```

Where the `in` is the underscored class name of the passed in  types

Reply:
`Expected Avram::InvalidSaveOperationError but nothing was raised`

not if I put the validation direct on the method `Expected Avram::InvalidSaveOperationError but nothing was raised`

ps: It's executing this method on the callback

Reply:
Hmm try this instead:

```crystal
def prepare
  previous_def
  validate_inclusion #...
end
```

I'm redoing how before_save works, but it is not in yet, so doing it in prepare might be better

Reply:
not even: `{{ type_declaration.var }}_type.add_error "test error"` raised exception hehe

Reply:
Hmm maybe need to add `class {{ @type }}::SaveOperation` also you can rebase and try `before_save` again. The new behavior has just been merged. If that still doesn't work I can pull it down and try to figure out what's up :P

Reply:
solved! 🎉